### PR TITLE
nodeset compiler correction and multiple dataset message read

### PR DESCRIPTION
**Pull request by ReidJunLi-Rockwell**
correction for below issues
https://github.com/open62541/open62541/issues/5475
https://github.com/open62541/open62541/issues/5473
https://github.com/open62541/open62541/issues/5472
https://github.com/open62541/open62541/issues/5469

### CODE REVIEW

File: src/pubsub/ua_pubsub_reader.c [hunk 1]
```diff
@@ -373,7 +373,14 @@ checkReaderIdentifier(UA_Server *server, UA_NetworkMessage *msg,
             }
         }
         if(msg->payloadHeaderEnabled) {
-            if(reader->config.dataSetWriterId != *msg->payloadHeader.dataSetPayloadHeader.dataSetWriterIds) {
+            UA_Byte anzDataSets = msg->payloadHeader.dataSetPayloadHeader.count;
```
Comment (jpfr) [inline src/pubsub/ua_pubsub_reader.c:376]: Is "anz" for the German "Anzahl"?
Keep to English also for variable names.

File: src/pubsub/ua_pubsub_reader.c [hunk 2]
```diff
@@ -373,7 +373,14 @@ checkReaderIdentifier(UA_Server *server, UA_NetworkMessage *msg,
...
+            UA_Byte iterator = 0;
+            for(iterator = 0; iterator < anzDataSets; iterator++) { 
+                if(reader->config.dataSetWriterId == *(msg->payloadHeader.dataSetPayloadHeader.dataSetWriterIds+iterator)) {
```
Comment (jpfr) [inline src/pubsub/ua_pubsub_reader.c:379]: Use `msg->payloadHeader.dataSetPayloadHeader.dataSetWriterIds[iterator]`.

File: src/pubsub/ua_pubsub_reader.c [hunk 3]
```diff
@@ -1321,10 +1328,14 @@ processMessageWithReader(UA_Server *server, UA_ReaderGroup *readerGroup,
     if(msg->payloadHeaderEnabled)
         totalDataSets = msg->payloadHeader.dataSetPayloadHeader.count;
     for(UA_Byte i = 0; i < totalDataSets; i++) {
-        UA_LOG_DEBUG_READER(&server->config.logger, reader,
+        /* map dataset reader to dataset message since multiple dataset reader may read this network message
+           otherwise the dataset message may be written to the wrong dataset reader.  */
+        if (reader->config.dataSetWriterId == *(msg->payloadHeader.dataSetPayloadHeader.dataSetWriterIds + i)) {
```
Comment (jpfr) [inline src/pubsub/ua_pubsub_reader.c:1333]: msg->payloadHeader.dataSetPayloadHeader.dataSetWriterIds[i]

File: tools/nodeset_compiler/nodeset_compiler.py [hunk 4]
```diff
@@ -114,12 +114,6 @@
 nsCount = 0
 loadedFiles = list()
 
-# Remove duplicate entries in typesArray, which occur when a nodeset has more than one depend.
-tmp_list = []
```
Comment (jpfr) [inline tools/nodeset_compiler/nodeset_compiler.py:118]: See my comment in #5471.
Let's deal with this in a separate PR. So we can merge the first batch fast.

**Review by jpfr — COMMENTED** [review]
Nice. It's good to see that issues that are immediately solved. 👍 
A few comments inline.

Please also use our convention for git commit messages.
In this case with the prefix `fix(pubsub): ` and `fix(nc): `.

Keep up the good work!

File: src/pubsub/ua_pubsub_reader.c (hunk 1 above)
Comment (ReidJunLi-Rockwell) [reply src/pubsub/ua_pubsub_reader.c]: changed to 'totalDataSets'

File: src/pubsub/ua_pubsub_reader.c (hunk 2 above)
Comment (ReidJunLi-Rockwell) [reply src/pubsub/ua_pubsub_reader.c]: updated

File: src/pubsub/ua_pubsub_reader.c (hunk 3 above)
Comment (ReidJunLi-Rockwell) [reply src/pubsub/ua_pubsub_reader.c]: updated

File: tools/nodeset_compiler/nodeset_compiler.py (hunk 4 above)
Comment (ReidJunLi-Rockwell) [reply tools/nodeset_compiler/nodeset_compiler.py:118]: reverted this file back

**Comment by ReidJunLi-Rockwell** [comment]
have cleaned up the history of my update, is it OK for merge?

**Comment by jpfr** [comment]
Your commits are merged via  #5487.

**Pull request closed without merging**